# WE RATE DOGS PROJECT

## Gather 

### Needed packages

In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt
import requests 
import json
import os.path
from os import path
import tweepy
import re
pd.set_option('display.max_colwidth', -1)

### Download `image_prediction`

In [2]:
if not path.exists("image-predictions.tsv"):
    url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'

    response = requests.get(url)
    with open(url.split("/")[-1],mode="wb") as file:
        file.write(response.content)

### Download from twitter

In [3]:
if not path.exists("tweet_json.txt"):
    consumer_key = 'skFJTei0s8uG9C74hajDurbfF'
    consumer_secret = 'olJzp2Xm43kY8LoajsmB1XRB1mIOhWoMSCjtCCJi0eNcId5jKA'
    access_token = '1234231320412327936-dJGZR40TDyztDs8YPyY4qSrbwqcLAv'
    access_secret = 'tI9Egkot9BGk1Jc0dXGqGuHLa7du1woAyw0PkoXpaoG7x'

    auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
    auth.set_access_token(access_token, access_secret)

    api = tweepy.API(auth,wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

    tweet_errors = {}

    for tweet_id in twee_df.tweet_id:
        try:
            tweet = api.get_status(tweet_id,tweet_mode='extended')
            print(tweet_id)
            with open("tweet_json.txt","a") as file:
                file.write(json.dumps(tweet._json))
                file.write("\r\n")
        except Exception as e:
            print(str(tweet_id) + ": " + str(e))
            tweet_errors[str(tweet_id)] = str(e)

### Load tweet_json.txt 

In [4]:
twee_data = []

with open('tweet_json.txt') as file:
    for line in file:
        twee_dict = json.loads(line)
        tweet_id = twee_dict['id']
        retweet_count = twee_dict['retweet_count']
        favorite_count = twee_dict['favorite_count']
        full_text = twee_dict['full_text']
        twee_data.append({'tweet_id':int(tweet_id),
                          'retweet_count' : int(retweet_count),
                          'favorite_count': int(favorite_count),
                          'text':full_text,
                          'favorited':twee_dict['favorited'],
                          'retweeted':twee_dict['retweeted']})
 
twee_data = pd.DataFrame(twee_data)

### Load `twitter-archive-enhanced.csv`

In [5]:
twee_arx = pd.read_csv('twitter-archive-enhanced.csv')

### Load `image_prediction.tsv`

In [6]:
image_pred = pd.read_csv("image-predictions.tsv",sep="\t")

## Assess

In [7]:
twee_data

favorite_count  favorited  retweet_count  retweeted  \
0     36288           False      7723           False       
1     31289           False      5714           False       
2     23558           False      3786           False       
3     39579           False      7882           False       
4     37802           False      8500           False       
5     19041           False      2849           False       
6     11088           False      1854           False       
7     61145           False      17238          False       
8     26197           False      3922           False       
9     29935           False      6714           False       
10    28818           False      6699           False       
11    26213           False      4560           False       
12    45136           False      9150           False       
13    25408           False      4103           False       
14    14212           False      2061           False       
15    23674           False      4870           False       
16    27358           False      4114           False       
17    24040           False      3878           False       
18    18560           False      3183           False       
19    20468           False      3165           False       
20    28419           False      4918           False       
21    43593           False      10764          False       
22    64603           False      16451          False       
23    31629           False      9563           False       
24    28777           False      5450           False       
25    32835           False      7015           False       
26    11266           False      2938           False       
27    21114           False      4086           False       
28    19872           False      2895           False       
29    111             False      4              False       
...   ...               ...     ..                ...       
2300  412             False      301            False       
2301  100             False      32             False       
2302  159             False      76             False       
2303  174             False      84             False       
2304  727             False      520            False       
2305  199             False      66             False       
2306  275             False      128            False       
2307  182             False      84             False       
2308  469             False      321            False       
2309  138             False      62             False       
2310  161             False      73             False       
2311  96              False      32             False       
2312  13685           False      6002           False       
2313  72              False      11             False       
2314  143             False      60             False       
2315  154             False      67             False       
2316  103             False      41             False       
2317  297             False      147            False       
2318  136             False      53             False       
2319  446             False      198            False       
2320  105             False      54             False       
2321  270             False      127            False       
2322  410             False      221            False       
2323  1129            False      778            False       
2324  125             False      51             False       
2325  96              False      39             False       
2326  272             False      132            False       
2327  112             False      41             False       
2328  121             False      42             False       
2329  2417            False      462            False       

                                                                                                                                                                        text  \
0     This is Phineas. He's a mystical boy. Onl

In [8]:
twee_arx

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193 NaN                    NaN                    
1     892177421306343426 NaN                    NaN                    
2     891815181378084864 NaN                    NaN                    
3     891689557279858688 NaN                    NaN                    
4     891327558926688256 NaN                    NaN                    
5     891087950875897856 NaN                    NaN                    
6     890971913173991426 NaN                    NaN                    
7     890729181411237888 NaN                    NaN                    
8     890609185150312448 NaN                    NaN                    
9     890240255349198849 NaN                    NaN                    
10    890006608113172480 NaN                    NaN                    
11    889880896479866881 NaN                    NaN                    
12    889665388333682689 NaN                    NaN                    
13    889638837579907072 NaN                    NaN                    
14    889531135344209921 NaN                    NaN                    
15    889278841981685760 NaN                    NaN                    
16    888917238123831296 NaN                    NaN                    
17    888804989199671297 NaN                    NaN                    
18    888554962724278272 NaN                    NaN                    
19    888202515573088257 NaN                    NaN                    
20    888078434458587136 NaN                    NaN                    
21    887705289381826560 NaN                    NaN                    
22    887517139158093824 NaN                    NaN                    
23    887473957103951883 NaN                    NaN                    
24    887343217045368832 NaN                    NaN                    
25    887101392804085760 NaN                    NaN                    
26    886983233522544640 NaN                    NaN                    
27    886736880519319552 NaN                    NaN                    
28    886680336477933568 NaN                    NaN                    
29    886366144734445568 NaN                    NaN                    
...                  ...  ..                     ..                    
2326  666411507551481857 NaN                    NaN                    
2327  666407126856765440 NaN                    NaN                    
2328  666396247373291520 NaN                    NaN                    
2329  666373753744588802 NaN                    NaN                    
2330  666362758909284353 NaN                    NaN                    
2331  666353288456101888 NaN                    NaN                    
2332  666345417576210432 NaN                    NaN                    
2333  666337882303524864 NaN                    NaN                    
2334  666293911632134144 NaN                    NaN                    
2335  666287406224695296 NaN                    NaN                    
2336  666273097616637952 NaN                    NaN                    
2337  666268910803644416 NaN                    NaN                    
2338  666104133288665088 NaN                    NaN                    
2339  666102155909144576 NaN                    NaN                    
2340  666099513787052032 NaN                    NaN                    
2341  666094000022159362 NaN                    NaN                    
2342  666082916733198337 NaN                    NaN                    
2343  666073100786774016 NaN                    NaN                    
2344  666071193221509120 NaN                    NaN                    
2345  666063827256086533 NaN                    NaN                    
2346  666058600524156928 NaN                    NaN                    
2347  666057090499244032 NaN                    NaN                    
2348  666055525042405380 NaN                    NaN                    
2349  666051853826850816 NaN                    NaN     

In [9]:
image_pred

tweet_id  \
0     666020888022790149   
1     666029285002620928   
2     666033412701032449   
3     666044226329800704   
4     666049248165822465   
5     666050758794694657   
6     666051853826850816   
7     666055525042405380   
8     666057090499244032   
9     666058600524156928   
10    666063827256086533   
11    666071193221509120   
12    666073100786774016   
13    666082916733198337   
14    666094000022159362   
15    666099513787052032   
16    666102155909144576   
17    666104133288665088   
18    666268910803644416   
19    666273097616637952   
20    666287406224695296   
21    666293911632134144   
22    666337882303524864   
23    666345417576210432   
24    666353288456101888   
25    666362758909284353   
26    666373753744588802   
27    666396247373291520   
28    666407126856765440   
29    666411507551481857   
...                  ...   
2045  886366144734445568   
2046  886680336477933568   
2047  886736880519319552   
2048  886983233522544640   
2049  887101392804085760   
2050  887343217045368832   
2051  887473957103951883   
2052  887517139158093824   
2053  887705289381826560   
2054  888078434458587136   
2055  888202515573088257   
2056  888554962724278272   
2057  888804989199671297   
2058  888917238123831296   
2059  889278841981685760   
2060  889531135344209921   
2061  889638837579907072   
2062  889665388333682689   
2063  889880896479866881   
2064  890006608113172480   
2065  890240255349198849   
2066  890609185150312448   
2067  890729181411237888   
2068  890971913173991426   
2069  891087950875897856   
2070  891327558926688256   
2071  891689557279858688   
2072  891815181378084864   
2073  892177421306343426   
2074  892420643555336193   

                                                                                      jpg_url  \
0     https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg                                           
1     https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg                                           
2     https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg                                           
3     https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg                                           
4     https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg                                           
5     https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg                                           
6     https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg                                           
7     https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg                                           
8     https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg                                           
9     https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg                                           
10    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg                                           
11    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg                                           
12    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg                                           
13    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg                                           
14    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg                                           
15    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg                                           
16    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg                                           
17    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg                                           
18    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg                                           
19    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg                                           
20    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg                                           
21    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg                                           
22    https://pbs.twimg.com/media/CT9OwFIWEAMuRj

In [10]:
twee_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2330 entries, 0 to 2329
Data columns (total 6 columns):
favorite_count    2330 non-null int64
favorited         2330 non-null bool
retweet_count     2330 non-null int64
retweeted         2330 non-null bool
text              2330 non-null object
tweet_id          2330 non-null int64
dtypes: bool(2), int64(3), object(1)
memory usage: 77.4+ KB


In [11]:
twee_arx.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [12]:
image_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [13]:
twee_arx.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03  7.800000e+01           7.800000e+01          
mean   7.427716e+17  7.455079e+17           2.014171e+16          
std    6.856705e+16  7.582492e+16           1.252797e+17          
min    6.660209e+17  6.658147e+17           1.185634e+07          
25%    6.783989e+17  6.757419e+17           3.086374e+08          
50%    7.196279e+17  7.038708e+17           4.196984e+09          
75%    7.993373e+17  8.257804e+17           4.196984e+09          
max    8.924206e+17  8.862664e+17           8.405479e+17          

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count  1.810000e+02         1.810000e+02              2356.000000        
mean   7.720400e+17         1.241698e+16              13.126486          
std    6.236928e+16         9.599254e+16              45.876648          
min    6.661041e+17         7.832140e+05              0.000000           
25%    7.186315e+17         4.196984e+09              10.000000          
50%    7.804657e+17         4.196984e+09              11.000000          
75%    8.203146e+17         4.196984e+09              12.000000          
max    8.874740e+17         7.874618e+17              1776.000000        

       rating_denominator  
count  2356.000000         
mean   10.455433           
std    6.745237            
min    0.000000            
25%    10.000000           
50%    10.000000           
75%    10.000000           
max    170.000000

In [14]:
twee_arx[twee_arx.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [15]:
twee_arx.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1637  684195085588783105 NaN                    NaN                    
910   757611664640446465 NaN                    NaN                    
302   836648853927522308 NaN                    NaN                    
1965  673345638550134785 NaN                    NaN                    
101   872668790621863937 NaN                    NaN                    

                      timestamp  \
1637  2016-01-05 02:09:54 +0000   
910   2016-07-25 16:21:11 +0000   
302   2017-02-28 18:46:45 +0000   
1965  2015-12-06 03:38:05 +0000   
101   2017-06-08 04:17:07 +0000   

                                                                                  source  \
1637  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
910   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
302   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1965  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
101   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                              text  \
1637  This is Tino. He really likes corndogs. 9/10 https://t.co/cUxGtnBfc2                                                                           
910   This is Cooper. He tries to come across as feisty but it never works for very long. 12/10 https://t.co/AVks8DjHwB                              
302   RT @SchafeBacon2016: @dog_rates Slightly disturbed by the outright profanity, but confident doggos were involved. 11/10, would tailgate aga…   
1965  This is Gerald. He's a fluffy lil yellow pup. Always looks like his favorite team just lost on a hail mary. 7/10 https://t.co/GpSkpN8kXS       
101   RT @loganamnosis: Penelope here is doing me quite a divertir. Well done, @dog_rates! Loving the pupdate. 14/10, je jouerais de nouveau. htt…   

      retweeted_status_id  retweeted_status_user_id  \
1637 NaN                  NaN                         
910  NaN                  NaN                         
302   8.366481e+17         7.124572e+17               
1965 NaN                  NaN                         
101   8.726576e+17         1.547674e+08               

     retweeted_status_timestamp  \
1637  NaN                         
910   NaN                         
302   2017-02-28 18:43:57 +0000   
1965  NaN                         
101   2017-06-08 03:32:35 +0000   

                                                                                                                                                                                        expanded_urls  \
1637  https://twitter.com/dog_rates/status/684195085588783105/photo/1                                                                                                                                   
910   https://twitter.com/dog_rates/status/757611664640446465/photo/1,https://twitter.com/dog_rates/status/757611664640446465/photo/1,https://twitter.com/dog_rates/status/757611664640446465/photo/1   
302   https://twitter.com/SchafeBacon2016/status/836648149003485187/photo/1                                                                                                                             
1965  https://twitter.com/dog_rates/status/673345638550134785/photo/1                                                                                                                                   
101   https://twitter.com/loganamnosis/status/872657584259551233/photo/1                                                                                                                                

      rating_numerator  rating_denominator    name doggo floofer pupper puppo  
1637  9                 10                  Tino    None  None    None   None  
910   12 

In [16]:
twee_arx.name.unique()

array(['Phineas', 'Tilly', 'Archie', 'Darla', 'Franklin', 'None', 'Jax',
       'Zoey', 'Cassie', 'Koda', 'Bruno', 'Ted', 'Stuart', 'Oliver',
       'Jim', 'Zeke', 'Ralphus', 'Canela', 'Gerald', 'Jeffrey', 'such',
       'Maya', 'Mingus', 'Derek', 'Roscoe', 'Waffles', 'Jimbo', 'Maisey',
       'Lilly', 'Earl', 'Lola', 'Kevin', 'Yogi', 'Noah', 'Bella',
       'Grizzwald', 'Rusty', 'Gus', 'Stanley', 'Alfy', 'Koko', 'Rey',
       'Gary', 'a', 'Elliot', 'Louis', 'Jesse', 'Romeo', 'Bailey',
       'Duddles', 'Jack', 'Emmy', 'Steven', 'Beau', 'Snoopy', 'Shadow',
       'Terrance', 'Aja', 'Penny', 'Dante', 'Nelly', 'Ginger', 'Benedict',
       'Venti', 'Goose', 'Nugget', 'Cash', 'Coco', 'Jed', 'Sebastian',
       'Walter', 'Sierra', 'Monkey', 'Harry', 'Kody', 'Lassie', 'Rover',
       'Napolean', 'Dawn', 'Boomer', 'Cody', 'Rumble', 'Clifford',
       'quite', 'Dewey', 'Scout', 'Gizmo', 'Cooper', 'Harold', 'Shikha',
       'Jamesy', 'Lili', 'Sammy', 'Meatball', 'Paisley', 'Albus',
       'Nept

In [17]:
twee_arx.rating_numerator.unique()

array([  13,   12,   14,    5,   17,   11,   10,  420,  666,    6,   15,
        182,  960,    0,   75,    7,   84,    9,   24,    8,    1,   27,
          3,    4,  165, 1776,  204,   50,   99,   80,   45,   60,   44,
        143,  121,   20,   26,    2,  144,   88])

In [18]:
twee_arx.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7       55 
14      54 
5       37 
6       32 
3       19 
4       17 
1       9  
2       9  
420     2  
0       2  
15      2  
75      2  
80      1  
20      1  
24      1  
26      1  
44      1  
50      1  
60      1  
165     1  
84      1  
88      1  
144     1  
182     1  
143     1  
666     1  
960     1  
1776    1  
17      1  
27      1  
45      1  
99      1  
121     1  
204     1  
Name: rating_numerator, dtype: int64

In [19]:
twee_arx.rating_denominator.unique()

array([ 10,   0,  15,  70,   7,  11, 150, 170,  20,  50,  90,  80,  40,
       130, 110,  16, 120,   2])

In [20]:
twee_arx.rating_denominator.value_counts()

10     2333
11     3   
50     3   
80     2   
20     2   
2      1   
16     1   
40     1   
70     1   
15     1   
90     1   
110    1   
120    1   
130    1   
150    1   
170    1   
7      1   
0      1   
Name: rating_denominator, dtype: int64

In [21]:
twee_arx.doggo.value_counts()

None     2259
doggo    97  
Name: doggo, dtype: int64

In [22]:
twee_arx.floofer.value_counts()

None       2346
floofer    10  
Name: floofer, dtype: int64

In [23]:
twee_arx.pupper.value_counts()

None      2099
pupper    257 
Name: pupper, dtype: int64

In [24]:
twee_arx.puppo.value_counts()

None     2326
puppo    30  
Name: puppo, dtype: int64

In [25]:
((twee_arx.doggo=='None') & (twee_arx.floofer=='None') & (twee_arx.pupper=='None') & (twee_arx.puppo=='None')).value_counts()

True     1976
False    380 
dtype: int64

In [26]:
(twee_arx.puppo=='puppo').sum()+(twee_arx.pupper=='pupper').sum()+(twee_arx.doggo=='doggo').sum()+(twee_arx.floofer=='floofer').sum()

394

#### Quality
##### *`twee_data`* table
- Missing records (2330 instead of 2356)

##### *`twee_arx`* table
- Missing data in columns `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id`, `retweeted_status_timestamp`, `expanded_urls`
- `timestamp` should be in Datetime datatype not string
- `retweeted_status_timestamp` should be in Datetime datatype not string

- Errors in `rating_numerator` and `rating_denominator` 

- Errors in `name`: such, a, not, one, my, his, this, unacceptable, .... They are all in lower case.
- stage `doggo` contains values None and doggo. They should be `0` for not doggo and `1` for being a doggo.
- Missing a lot of data in stage variable (doggo, floofer, pupper, puppo) <span style="color:blue">(can't clean)</span>

- Unnecessary columns `source`, `timestamp`, `expanded_urls`

##### *`image_pred`* table
- Missing records (2075 instead of 2356)
- Dog's breeds are lower case and sometime upper case.

#### Tidiness

- One varible in four columns in `twee_arx` table (`doggo`,`floofer`,`pupper`,`puppo`)
- Columns `retweet_count` and `favorite_count` should also be part of the `twee_arx` table
- column `text` in `twee_arx` duplicated in `twee_data` table
- `text` column in `twee_data` table should be split into `name`, `rating_numerator` and `rating_denominator`
- `twee_data` and `twee_arx` contain retweet record. These record should be remove because they are not part of our analysis.

## Clean

In [27]:
twee_arx_clean = twee_arx.copy()
twee_data_clean = twee_data.copy()
image_pred_clean = image_pred.copy()

### Missing data

#### Missing records

- `twee_data`: Missing records (2330 instead of 2356). These happen due to the deleted status of the tweets so it can't be cleaned. We will cut off the deleted tweets from other tables.
- `image_pred`: Missing records (2075 instead of 2356). This issue can't be cleaned since we don't have the prediction model. So the same solution as the missing data in `twee_data` table. We will created the intersection of all tables.

#### Define:
Find the tweets that are all three tables and cut off the other tweets from the tables.

#### Code:

In [28]:
list_tweets = list(set(twee_data.tweet_id) & set(twee_arx.tweet_id) & set(image_pred.tweet_id))

In [29]:
twee_arx_clean.set_index('tweet_id',inplace=True)
twee_data_clean.set_index('tweet_id',inplace=True)
image_pred_clean.set_index('tweet_id',inplace=True)
twee_arx_clean = twee_arx_clean.loc[list_tweets]
twee_data_clean = twee_data_clean.loc[list_tweets]
image_pred_clean = image_pred_clean.loc[list_tweets]

#### Test

In [30]:
set(twee_data_clean.index) == set(twee_arx_clean.index) == set(image_pred_clean.index)

True

#### Missing data in columns `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id` in `twee_arx` table

- Actually, these are not the missing data issues. These are poorly represented values. Those `NaN` mean the tweet is not a reply in another tweet. Since we are only interested in whether a tweet is an original tweet or a reply, we should use 0 for a reply and 1 for an orignal tweet.

#### Define:

In `twee_arx` table:
-  Create a new column name original_tweet which has value: 

    - `True` if the corresponding value in both columns `in_reply_to_status_id` and `retweeted_status_id` is `NaN`
    - `False` otherwise
    
- Delete unnecessary columns `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id`, `retweeted_status_timestamp`, `expanded_urls`.

#### Code:

In [31]:
twee_arx_clean['original_tweet'] = twee_arx_clean.in_reply_to_status_id.isna() & twee_arx_clean.retweeted_status_id.isna()
twee_arx_clean.drop(['in_reply_to_status_id','in_reply_to_user_id', 'retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp'],axis=1,inplace=True)

#### Test

In [32]:
twee_arx_clean.sample(5)

timestamp  \
tweet_id                                        
874296783580663808  2017-06-12 16:06:11 +0000   
693095443459342336  2016-01-29 15:36:45 +0000   
668221241640230912  2015-11-22 00:15:33 +0000   
677687604918272002  2015-12-18 03:11:30 +0000   
670842764863651840  2015-11-29 05:52:33 +0000   

                                                                                                source  \
tweet_id                                                                                                 
874296783580663808  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
693095443459342336  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
668221241640230912  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
677687604918272002  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
670842764863651840  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                               text  \
tweet_id                                                                                                                                                              
874296783580663808  This is Jed. He may be the fanciest pupper in the game right now. Knows it too. 13/10 would sign modeling contract https://t.co/0YplNnSMEm        
693095443459342336  This is Milo. He doesn't understand your fancy human gestures. Will lick instead. 10/10 can't faze this pupper https://t.co/OhodPIDOpW            
668221241640230912  These two dogs are Bo &amp; Smittens. Smittens is trying out a new deodorant and wanted Bo to smell it. 10/10 true pals https://t.co/4pw1QQ6udh   
677687604918272002  This was Cindy's face when she heard Susan forgot the snacks for after the kid's soccer game. 11/10 https://t.co/gzkuVGRgAD                       
670842764863651840  After so many requests... here you go.\n\nGood dogg. 420/10 https://t.co/yfAAo1gdeY                                                               

                                                                      expanded_urls  \
tweet_id                                                                              
874296783580663808  https://twitter.com/dog_rates/status/874296783580663808/photo/1   
693095443459342336  https://twitter.com/dog_rates/status/693095443459342336/photo/1   
668221241640230912  https://twitter.com/dog_rates/status/668221241640230912/photo/1   
677687604918272002  https://twitter.com/dog_rates/status/677687604918272002/photo/1   
670842764863651840  https://twitter.com/dog_rates/status/670842764863651840/photo/1   

                    rating_numerator  rating_denominator  name doggo floofer  \
tweet_id                                                                       
874296783580663808  13                10                  Jed   None  None     
693095443459342336  10                10                  Milo  None  None     
668221241640230912  10                10                  None  None  None     
677687604918272002  11                10                  None  None  None     
670842764863651840  420               10                  None  None  None     

                    pupper puppo  original_tweet  
tweet_id                                          
874296783580663808  pupper  None  True            
693095443459342336  pupper  None  True            
668221241640230912  None    None  True            
677687604918272002  None    None  True            
670842764863651840  None    None  True

In [33]:
twee_arx_clean.original_tweet.value_counts()

True     1963
False    95  
Name: original_tweet, dtype: int64

### None values

#### Define:

Change all None value to NaN using `np.replace`

#### Code:

In [34]:
twee_arx_clean.replace('None',np.NaN,inplace=True)

#### Test:

In [35]:
twee_arx_clean.head()

timestamp  \
tweet_id                                        
666051853826850816  2015-11-16 00:35:11 +0000   
666049248165822465  2015-11-16 00:24:50 +0000   
666094000022159362  2015-11-16 03:22:39 +0000   
666373753744588802  2015-11-16 21:54:18 +0000   
666055525042405380  2015-11-16 00:49:46 +0000   

                                                                                                source  \
tweet_id                                                                                                 
666051853826850816  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
666049248165822465  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
666094000022159362  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
666373753744588802  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
666055525042405380  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                            text  \
tweet_id                                                                                                                                                           
666051853826850816  This is an odd dog. Hard on the outside but loving on the inside. Petting still fun. Doesn't play catch well. 2/10 https://t.co/v5A4vzSDdc     
666049248165822465  Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq                       
666094000022159362  This appears to be a Mongolian Presbyterian mix. Very tired. Tongue slip confirmed. 9/10 would lie down with https://t.co/mnioXo3IfP           
666373753744588802  Those are sunglasses and a jean jacket. 11/10 dog cool af https://t.co/uHXrPkUEyl                                                              
666055525042405380  Here is a Siberian heavily armored polar bear mix. Strong owner. 10/10 I would do unspeakable things to pet this dog https://t.co/rdivxLiqEt   

                                                                      expanded_urls  \
tweet_id                                                                              
666051853826850816  https://twitter.com/dog_rates/status/666051853826850816/photo/1   
666049248165822465  https://twitter.com/dog_rates/status/666049248165822465/photo/1   
666094000022159362  https://twitter.com/dog_rates/status/666094000022159362/photo/1   
666373753744588802  https://twitter.com/dog_rates/status/666373753744588802/photo/1   
666055525042405380  https://twitter.com/dog_rates/status/666055525042405380/photo/1   

                    rating_numerator  rating_denominator name doggo floofer  \
tweet_id                                                                      
666051853826850816  2                 10                  an   NaN   NaN      
666049248165822465  5                 10                  NaN  NaN   NaN      
666094000022159362  9                 10                  NaN  NaN   NaN      
666373753744588802  11                10                  NaN  NaN   NaN      
666055525042405380  10                10                  a    NaN   NaN      

                   pupper puppo  original_tweet  
tweet_id                                         
666051853826850816  NaN    NaN   True            
666049248165822465  NaN    NaN   True            
666094000022159362  NaN    NaN   True            
666373753744588802  NaN    NaN   True            
666055525042405380  NaN    NaN   True

### Unnecessary columns

#### Define:

Remove the unneccessary columns `source`, `expanded_urls`

#### Code:

In [36]:
twee_arx_clean.drop(['source','expanded_urls'],axis=1,inplace=True)

#### Test:

In [37]:
twee_arx_clean.sample(5)

timestamp  \
tweet_id                                        
816829038950027264  2017-01-05 02:09:53 +0000   
773308824254029826  2016-09-06 23:56:05 +0000   
759099523532779520  2016-07-29 18:53:24 +0000   
741793263812808706  2016-06-12 00:44:30 +0000   
678255464182861824  2015-12-19 16:47:58 +0000   

                                                                                                                                                            text  \
tweet_id                                                                                                                                                           
816829038950027264  RT @dog_rates: This is Betty. She's assisting with the dishes. Such a good puppo. 12/10 h*ckin helpful af https://t.co/dgvTPZ9tgI              
773308824254029826  This is Bungalo. She uses that face to get what she wants. It works unbelievably well. 12/10 would never say no to https://t.co/0Fcft7jl4N     
759099523532779520  Meet Toby. He has a drinking problem. Inflatable marijuana plant in the back is also not a good look. 7/10 cmon Toby https://t.co/Cim4DSj6Oi   
741793263812808706  When your crush won't pay attention to you. Both 10/10 tragic af https://t.co/d3LELGVlqu                                                       
678255464182861824  This is Jessifer. She is a Bismoth Teriyaki. Flowers being attacked by hurricanes on bandana (rad). 9/10 stellar pup https://t.co/nZhmRwZzWv   

                    rating_numerator  rating_denominator      name doggo  \
tweet_id                                                                   
816829038950027264  12                10                  Betty     NaN    
773308824254029826  12                10                  Bungalo   NaN    
759099523532779520  7                 10                  Toby      NaN    
741793263812808706  10                10                  NaN       NaN    
678255464182861824  9                 10                  Jessifer  NaN    

                   floofer pupper  puppo  original_tweet  
tweet_id                                                  
816829038950027264  NaN     NaN    puppo  False           
773308824254029826  NaN     NaN    NaN    True            
759099523532779520  NaN     NaN    NaN    True            
741793263812808706  NaN     NaN    NaN    True            
678255464182861824  NaN     NaN    NaN    True

### Tidiness

#### One varible in four columns in *`twee_arx`* table (doggo,floofer,pupper,puppo)

#### Define:

- Define a function `get_stage` which input a tweet_id and return the stage of the dog based on columns *doggo, floofer, pupper* and *puppo*

- Create a new column named `stage` using `get_stage` function

- Remove the columns *doggo, floofer, pupper* and *puppo*

#### Code:

In [38]:
def get_stage(x):
    if x.doggo=='doggo':
        return 'doggo'
    elif x.floofer=='floofer':
        return 'floofer'
    elif x.pupper=='pupper':
        return 'pupper'
    elif x.puppo=='puppo':
        return 'puppo'
    else:
        return np.nan


In [39]:
twee_arx_clean['stage'] = twee_arx_clean.apply(get_stage,axis=1)
twee_arx_clean.drop(['doggo','floofer','pupper','puppo'],axis=1,inplace=True)

#### Test:

In [40]:
list(twee_arx_clean)

['timestamp',
 'text',
 'rating_numerator',
 'rating_denominator',
 'name',
 'original_tweet',
 'stage']

#### Columns `retweet_count` and `favorite_count` should also be part of the `twee_arx` table

#### Define:

Merge the *retweet_count* and *favorite_count* columns to the `twee_arx` table, joining on *tweet_id*

#### Code:

In [41]:
twee_arx_clean.reset_index()
twee_arx_clean = pd.merge(twee_arx_clean, twee_data_clean[['retweet_count','favorite_count']],
                            on=['tweet_id'], how='left')

#### Test:

In [42]:
twee_arx_clean.head()

timestamp  \
tweet_id                                        
666051853826850816  2015-11-16 00:35:11 +0000   
666049248165822465  2015-11-16 00:24:50 +0000   
666094000022159362  2015-11-16 03:22:39 +0000   
666373753744588802  2015-11-16 21:54:18 +0000   
666055525042405380  2015-11-16 00:49:46 +0000   

                                                                                                                                                            text  \
tweet_id                                                                                                                                                           
666051853826850816  This is an odd dog. Hard on the outside but loving on the inside. Petting still fun. Doesn't play catch well. 2/10 https://t.co/v5A4vzSDdc     
666049248165822465  Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq                       
666094000022159362  This appears to be a Mongolian Presbyterian mix. Very tired. Tongue slip confirmed. 9/10 would lie down with https://t.co/mnioXo3IfP           
666373753744588802  Those are sunglasses and a jean jacket. 11/10 dog cool af https://t.co/uHXrPkUEyl                                                              
666055525042405380  Here is a Siberian heavily armored polar bear mix. Strong owner. 10/10 I would do unspeakable things to pet this dog https://t.co/rdivxLiqEt   

                    rating_numerator  rating_denominator name  original_tweet  \
tweet_id                                                                        
666051853826850816  2                 10                  an   True             
666049248165822465  5                 10                  NaN  True             
666094000022159362  9                 10                  NaN  True             
666373753744588802  11                10                  NaN  True             
666055525042405380  10                10                  a    True             

                   stage  retweet_count  favorite_count  
tweet_id                                                 
666051853826850816  NaN   778            1129            
666049248165822465  NaN   39             96              
666094000022159362  NaN   67             154             
666373753744588802  NaN   84             174             
666055525042405380  NaN   221            410

#### Prediction columns `p1`:`p_3_dog` should be part of `twee_arx` table

#### Define:

Merge the *p1* to *p3_dog* columns to the `twee_arx` table, joining on *tweet_id*

#### Code:

In [43]:
pred_cols = list(image_pred_clean.columns)[2:]

twee_arx_clean.reset_index()
twee_arx_clean = pd.merge(twee_arx_clean, image_pred_clean[pred_cols],
                            on=['tweet_id'], how='left')

#### Test:

In [44]:
twee_arx_clean

timestamp  \
tweet_id                                        
666051853826850816  2015-11-16 00:35:11 +0000   
666049248165822465  2015-11-16 00:24:50 +0000   
666094000022159362  2015-11-16 03:22:39 +0000   
666373753744588802  2015-11-16 21:54:18 +0000   
666055525042405380  2015-11-16 00:49:46 +0000   
667090893657276420  2015-11-18 21:23:57 +0000   
667878741721415682  2015-11-21 01:34:35 +0000   
670046952931721218  2015-11-27 01:10:17 +0000   
671115716440031232  2015-11-29 23:57:10 +0000   
672594978741354496  2015-12-04 01:55:13 +0000   
700002074055016451  2016-02-17 17:01:14 +0000   
740365076218183684  2016-06-08 02:09:24 +0000   
745712589599014916  2016-06-22 20:18:30 +0000   
687818504314159109  2016-01-15 02:08:05 +0000   
695767669421768709  2016-02-06 00:35:13 +0000   
669567591774625800  2015-11-25 17:25:28 +0000   
676821958043033607  2015-12-15 17:51:44 +0000   
887473957103951883  2017-07-19 00:47:34 +0000   
668643542311546881  2015-11-23 04:13:37 +0000   
796080075804475393  2016-11-08 20:00:55 +0000   
741303864243200000  2016-06-10 16:19:48 +0000   
671789708968640512  2015-12-01 20:35:22 +0000   
673708611235921920  2015-12-07 03:40:24 +0000   
674042553264685056  2015-12-08 01:47:22 +0000   
798673117451325440  2016-11-15 23:44:44 +0000   
711306686208872448  2016-03-19 21:41:44 +0000   
716802964044845056  2016-04-04 01:41:58 +0000   
736225175608430592  2016-05-27 15:58:54 +0000   
738537504001953792  2016-06-03 01:07:16 +0000   
765222098633691136  2016-08-15 16:22:20 +0000   
...                                       ...   
891689557279858688  2017-07-30 15:58:51 +0000   
667435689202614272  2015-11-19 20:14:03 +0000   
818536468981415936  2017-01-09 19:14:36 +0000   
674019345211760640  2015-12-08 00:15:09 +0000   
680206703334408192  2015-12-25 02:01:30 +0000   
685268753634967552  2016-01-08 01:16:17 +0000   
685547936038666240  2016-01-08 19:45:39 +0000   
680440374763077632  2015-12-25 17:30:01 +0000   
825876512159186944  2017-01-30 01:21:19 +0000   
832397543355072512  2017-02-17 01:13:34 +0000   
816336735214911488  2017-01-03 17:33:39 +0000   
719332531645071360  2016-04-11 01:13:34 +0000   
718246886998687744  2016-04-08 01:19:36 +0000   
686003207160610816  2016-01-10 01:54:44 +0000   
691459709405118465  2016-01-25 03:16:56 +0000   
672640509974827008  2015-12-04 04:56:09 +0000   
713411074226274305  2016-03-25 17:03:49 +0000   
748307329658011649  2016-06-30 00:09:04 +0000   
819347104292290561  2017-01-12 00:55:47 +0000   
827653905312006145  2017-02-03 23:04:02 +0000   
828372645993398273  2017-02-05 22:40:03 +0000   
699413908797464576  2016-02-16 02:04:04 +0000   
828409743546925057  2017-02-06 01:07:28 +0000   
697881462549430272  2016-02-11 20:34:41 +0000   
709566166965075968  2016-03-15 02:25:31 +0000   
689661964914655233  2016-01-20 04:13:20 +0000   
669214165781868544  2015-11-24 18:01:05 +0000   
834209720923721728  2017-02-22 01:14:30 +0000   
673317986296586240  2015-12-06 01:48:12 +0000   
674790488185167872  2015-12-10 03:19:24 +0000   

                                                                                                                                                                         text  \
tweet_id                                                                                                                                                                        
666051853826850816  This is an odd dog. Hard on the outside but loving on the inside. Petting still fun. Doesn't play catch well. 2/10 https://t.co/v5A4vzSDdc                  
666049248165822465  Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq                                    
666094000022159362  This appears to be a Mongolian Presbyterian mix. Very tired. Tongue slip confirmed. 9/10 would lie down with https://t.co/mnioXo3IfP                        
666373753744588802  Those are sunglasses and a jean jacket. 11/

#### `twee_arx` contain retweeted reply record 

#### Define:

Remove the tweets which are retweets from other status or reply to others using the column `original_tweet` and then drop this column. Store the result to `twitter_archive` table which should have 1963 records.

#### Code:

In [45]:
twitter_archive = twee_arx_clean[twee_arx_clean.original_tweet==True].copy()
twitter_archive.drop(['original_tweet'],axis=1,inplace=True)

#### Test:

In [46]:
len(twitter_archive)

1963

### Quality

#### Errors in `name` column in `twee_arx` table

#### Define:

 Define a function named `correct_name` which inputs a row and check if the first character in it's name is whether upper case or lower case. 
 - If it is upper case then it's a correct name. Return the current name.
 - If it is lower case then it's a incorrect name. We will check if in the `text` contains the word `named` then after that word should be the correct name and return that name. Otherwise return `NaN` because the dog's name does not specify.
 
Apply `correct_name` function to the `twee_arx_clean` table.

#### Code:

In [47]:

def correct_name(row):
    if pd.isnull(row['name']):
        return row['name']
    elif row['name'][0].isupper():
        return row['name']
    else:
        words = row.text.split(" ")
        if "named" in words:
            return words[words.index("named")+1]
        else:
            return np.nan

twee_arx_clean['name'] = twee_arx_clean.apply(correct_name,axis=1)

#### Test:

In [48]:
twee_arx_clean[['text','name']]

text  \
tweet_id                                                                                                                                                                        
666051853826850816  This is an odd dog. Hard on the outside but loving on the inside. Petting still fun. Doesn't play catch well. 2/10 https://t.co/v5A4vzSDdc                  
666049248165822465  Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq                                    
666094000022159362  This appears to be a Mongolian Presbyterian mix. Very tired. Tongue slip confirmed. 9/10 would lie down with https://t.co/mnioXo3IfP                        
666373753744588802  Those are sunglasses and a jean jacket. 11/10 dog cool af https://t.co/uHXrPkUEyl                                                                           
666055525042405380  Here is a Siberian heavily armored polar bear mix. Strong owner. 10/10 I would do unspeakable things to pet this dog https://t.co/rdivxLiqEt                
667090893657276420  This is Clybe. He is an Anemone Valdez. One ear works. Can look in 2 different directions at once. Tongue slip. 7/10 https://t.co/Ks0jZtdIrr                
667878741721415682  This is Tedrick. He lives on the edge. Needs someone to hit the gas tho. Other than that he's a baller. 10&amp;2/10 https://t.co/LvP1TTYSCN                 
670046952931721218  This is Ben &amp; Carson. It's impossible for them to tilt their heads in the same direction. Cheeky wink by Ben. 11/10s https://t.co/465sIBdvzU            
671115716440031232  Meet Phred. He isn't steering, looking at the road, or wearing a seatbelt. Phred is a rolling tornado of danger 6/10 https://t.co/mZD7Bo7HfV                
672594978741354496  Meet Scott. Just trying to catch his train to work. Doesn't need everybody staring. 9/10 ignore the haters pupper https://t.co/jyXbZ35MYz                   
700002074055016451  This is Thumas. He covered himself in nanners for maximum camouflage. It didn't work. I can still see u Thumas. 9/10 https://t.co/x0ZDlNqfb1                
740365076218183684  When the photographer forgets to tell you where to look... 10/10 https://t.co/u1GHWxhC85                                                                    
745712589599014916  This is Percy. He fell asleep at the wheel. Irresponsible af. 7/10 absolute menace on the roadway https://t.co/QHbvtvaw8E                                   
687818504314159109  With great pupper comes great responsibility. 12/10 https://t.co/hK6xB042EP                                                                                 
695767669421768709  If you are aware of who is making these please let me know. 13/10 vroom vroom https://t.co/U0D1sbIDrG                                                       
669567591774625800  Meet Kollin. He's a Parakeetian Badminton from Denmark. Great artist. Taking break from research. Loves wicker 9/10 https://t.co/XPLB3eoXiX                 
676821958043033607  Finally some constructive political change in this country. 11/10 https://t.co/mvQaETHVSb                                                                   
887473957103951883  This is Canela. She attempted some fancy porch pics. They were unsuccessful. 13/10 someone help her https://t.co/cLyzpcUcMX                                 
668643542311546881  Fascinating dog here. Loves beach. Oddly long nose for dog. Massive ass paws. Hard to cuddle w. 3/10 would still pet https://t.co/IiSdmhkC5N                
796080075804475393  This is Yogi. He's 98% floof. Snuggable af. 12/10 https://t.co/opoXKxmfFm                                                                                   
741303864243200000  This is Theodore. He just saw an adult wearing crocs. Did him a frighten. Lost other ear back in nam. 12/10 hero af https://t.co/O4iw9NlLaQ                 
671789708968640512  This is space pup. He's very confused. Tries to moonwalk at one point. Super s

#### Column `timestamp` should be in Datetime datatype not string

#### Define:

Using `pd.datetime` to change the datatype of the column to `datetime`

#### Code:

In [49]:
twitter_archive.timestamp = pd.to_datetime(twitter_archive.timestamp)

#### Test:

In [50]:
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1963 entries, 666051853826850816 to 674790488185167872
Data columns (total 17 columns):
timestamp             1963 non-null datetime64[ns]
text                  1963 non-null object
rating_numerator      1963 non-null int64
rating_denominator    1963 non-null int64
name                  1440 non-null object
stage                 302 non-null object
retweet_count         1963 non-null int64
favorite_count        1963 non-null int64
p1                    1963 non-null object
p1_conf               1963 non-null float64
p1_dog                1963 non-null bool
p2                    1963 non-null object
p2_conf               1963 non-null float64
p2_dog                1963 non-null bool
p3                    1963 non-null object
p3_conf               1963 non-null float64
p3_dog                1963 non-null bool
dtypes: bool(3), datetime64[ns](1), float64(3), int64(4), object(6)
memory usage: 235.8+ KB


#### Stage `doggo` contains values None and doggo. They should be `0` for not doggo and `1` for being a doggo.

This has been fixed in the tidiness process already.

#### Errors in `rating_numerator` and `rating_denominator` 

This happens because there are more than one phrase $n/d$ in the `text`. Note that the convention of the rating should be n/10 but people usually make fun of it by using the from $(n*k)/(10*k)$. So let's assume that the base denominator is 10 then we only need the numerator as the rating. In this case, it is easier to dump the rating columns and create the new correct ones.

#### Define:

Define a function named `get_rating` which inputs a row and then
- split the `text` into words
- check for words contain the slash `/` and choose the last one
- for the chosen word containing `/`, split them using the separate character `/` then numerator is the first part and denominator is the second part
- divide the numerator and denominator by a common factor so that the denominator will become 10.

*Note to deal with something unsual like "10&2/10"*

Apply the function `get_rating` to the dataframe to get a new column named `rating`.

Drop the columns `rating_numerator` and `rating_denominator`.

#### Code:

In [84]:
def get_rating(row):
    #trim the last word that contains the link to the tweet
    text = row.text.split("http")[0]
    words = text.split(" ")
    
    rating = [w for w in words if w.find("/")>=0][-1]
    n,d = rating.split("/")[-2:]
    
    # Deal with unsually rating
    n = re.findall('\d+',n)[-1]
    n = int(n)
    d = re.findall("\d+",d)[0]
    d = int(d)
    # Divide out the common factor
    d = int(d)
    if d%10==0:
        return n//(d//10)
    else:
        return np.nan
twitter_archive['rating'] = twitter_archive.apply(get_rating,axis=1)

# Drop old rating columns
twitter_archive.drop(['rating_numerator','rating_denominator'],axis=1, inplace=True)

#### Test:

In [88]:
twitter_archive[['text','rating']].sample(50)

text  \
tweet_id                                                                                                                                                                              
715200624753819648  This is Michelangelope. He's half coffee cup. Rare af. 12/10 would hug until someone stopped me https://t.co/tvVDY0G911                                           
677918531514703872  "Everything looks pretty good in there. Make sure to brush your gums. Been flossing? How's school going?" Both 10/10 https://t.co/lWL2IMJqLR                      
673343217010679808  IT'S SO SMALL ERMERGERF 11/10 https://t.co/dNUbKOSiWW                                                                                                             
819711362133872643  This is Howie. He just bloomed. 11/10 revolutionary af https://t.co/m5fYxrO3IU                                                                                    
857263160327368704  This is Oscar and Oliver. Oliver shrunk Oscar. Oscar isn't pleased about it. Quite pupset tbh. Oliver doesn't seem to mind. Both 13/10 https://t.co/e3U4NReleC    
692905862751522816  This is Wally. He's being abducted by aliens. 10/10 poor pupper https://t.co/EiF659Bgjn                                                                           
688547210804498433  This is Frönq. He got caught stealing a waffle. Damn it Frönq. 9/10 https://t.co/7ycWCUrjmZ                                                                       
668268907921326080  Here we have an Azerbaijani Buttermilk named Guss. He sees a demon baby Hitler behind his owner. 10/10 stays alert https://t.co/aeZykWwiJN                        
731285275100512256  This is Solomon. He's a Beneroo Cumberflop. 12/10 would hug passionately https://t.co/5phLAnGPTP                                                                  
670385711116361728  Meet Larry. He's a Panoramic Benzoate. Can shoot lasers out of his eyes. Very neat. Stuck in that position tho. 8/10 https://t.co/MAZx8MPF0S                      
872122724285648897  This is Kody. He's a baller. Wishes he was a little bit taller. Double dribbles often. Still 12/10 would happily get dunked on https://t.co/PKSpmiefwN            
701981390485725185  This is Fiji. She's a Powdered Stegafloof. Very rare. 12/10 https://t.co/fZRob6eotY                                                                               
670815497391357952  This is Sage. He likes to burn shit. 10/10 https://t.co/nLYruSMRe6                                                                                                
739238157791694849  Here's a doggo blowing bubbles. It's downright legendary. 13/10 would watch on repeat forever (vid by Kent Duryee) https://t.co/YcXgHfp1EC                        
693590843962331137  Meet Phil. He's big af. Currently destroying this nice family home. Completely uncalled for. 3/10 not a good pupper https://t.co/fShNNhBWYx                       
868880397819494401  This is Walter. He won't start hydrotherapy without his favorite floatie. 14/10 keep it pup Walter https://t.co/r28jFx9uyF                                        
793241302385262592  This is Tucker. He's out here bustin h*ckin ghosts. 13/10 dedicated af https://t.co/Ap477GhwXt                                                                    
674632714662858753  Rare submerged pup here. Holds breath for a long time. Frowning because that spoon ignores him. 5/10 would still pet https://t.co/EJzzNHE8bE                      
791672322847637504  When she says you're a good boy and you know you're a good boy because you're a good boy. 13/10 https://t.co/O5IUmRHRIh                                           
696405997980676096  This is Berb. He just found out that they have made 31 Kidz Bop CD's. Downright terrifying. 7/10 hang in there Berb https://t.co/CIFLjiTFwZ                       
786709082849828864  This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wuqaPS 

In [90]:
twitter_archive

timestamp  \
tweet_id                                 
666051853826850816 2015-11-16 00:35:11   
666049248165822465 2015-11-16 00:24:50   
666094000022159362 2015-11-16 03:22:39   
666373753744588802 2015-11-16 21:54:18   
666055525042405380 2015-11-16 00:49:46   
667090893657276420 2015-11-18 21:23:57   
667878741721415682 2015-11-21 01:34:35   
670046952931721218 2015-11-27 01:10:17   
671115716440031232 2015-11-29 23:57:10   
672594978741354496 2015-12-04 01:55:13   
700002074055016451 2016-02-17 17:01:14   
740365076218183684 2016-06-08 02:09:24   
745712589599014916 2016-06-22 20:18:30   
687818504314159109 2016-01-15 02:08:05   
669567591774625800 2015-11-25 17:25:28   
676821958043033607 2015-12-15 17:51:44   
887473957103951883 2017-07-19 00:47:34   
668643542311546881 2015-11-23 04:13:37   
796080075804475393 2016-11-08 20:00:55   
741303864243200000 2016-06-10 16:19:48   
671789708968640512 2015-12-01 20:35:22   
673708611235921920 2015-12-07 03:40:24   
674042553264685056 2015-12-08 01:47:22   
711306686208872448 2016-03-19 21:41:44   
716802964044845056 2016-04-04 01:41:58   
736225175608430592 2016-05-27 15:58:54   
738537504001953792 2016-06-03 01:07:16   
765222098633691136 2016-08-15 16:22:20   
774314403806253056 2016-09-09 18:31:54   
691820333922455552 2016-01-26 03:09:55   
...                                ...   
891689557279858688 2017-07-30 15:58:51   
667435689202614272 2015-11-19 20:14:03   
818536468981415936 2017-01-09 19:14:36   
674019345211760640 2015-12-08 00:15:09   
680206703334408192 2015-12-25 02:01:30   
685268753634967552 2016-01-08 01:16:17   
685547936038666240 2016-01-08 19:45:39   
680440374763077632 2015-12-25 17:30:01   
825876512159186944 2017-01-30 01:21:19   
832397543355072512 2017-02-17 01:13:34   
816336735214911488 2017-01-03 17:33:39   
719332531645071360 2016-04-11 01:13:34   
718246886998687744 2016-04-08 01:19:36   
686003207160610816 2016-01-10 01:54:44   
691459709405118465 2016-01-25 03:16:56   
672640509974827008 2015-12-04 04:56:09   
713411074226274305 2016-03-25 17:03:49   
748307329658011649 2016-06-30 00:09:04   
819347104292290561 2017-01-12 00:55:47   
827653905312006145 2017-02-03 23:04:02   
828372645993398273 2017-02-05 22:40:03   
699413908797464576 2016-02-16 02:04:04   
828409743546925057 2017-02-06 01:07:28   
697881462549430272 2016-02-11 20:34:41   
709566166965075968 2016-03-15 02:25:31   
689661964914655233 2016-01-20 04:13:20   
669214165781868544 2015-11-24 18:01:05   
834209720923721728 2017-02-22 01:14:30   
673317986296586240 2015-12-06 01:48:12   
674790488185167872 2015-12-10 03:19:24   

                                                                                                                                                                         text  \
tweet_id                                                                                                                                                                        
666051853826850816  This is an odd dog. Hard on the outside but loving on the inside. Petting still fun. Doesn't play catch well. 2/10 https://t.co/v5A4vzSDdc                  
666049248165822465  Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq                                    
666094000022159362  This appears to be a Mongolian Presbyterian mix. Very tired. Tongue slip confirmed. 9/10 would lie down with https://t.co/mnioXo3IfP                        
666373753744588802  Those are sunglasses and a jean jacket. 11/10 dog cool af https://t.co/uHXrPkUEyl                                                                           
666055525042405380  Here is a Siberian heavily armored polar bear mix. Strong owner. 10/10 I would do unspeakable things to pet this dog https://t.co/rdivxLiqEt                
667090893657276420  This is Clybe. He is an Anemone Valdez. One ear works. Can look in 2 different directions at once. Tongue slip. 7/10 https:

#### Dog's breeds are lower case and sometime upper case.

#### Define:

Convert all dog's breeds in the prediction to lower case.

#### Code:

In [93]:
cols = ['p1','p2','p3']

for col in cols:
    twitter_archive[col] = twitter_archive[col].apply(lambda x: x.lower())

#### Test:

In [94]:
twitter_archive

timestamp  \
tweet_id                                 
666051853826850816 2015-11-16 00:35:11   
666049248165822465 2015-11-16 00:24:50   
666094000022159362 2015-11-16 03:22:39   
666373753744588802 2015-11-16 21:54:18   
666055525042405380 2015-11-16 00:49:46   
667090893657276420 2015-11-18 21:23:57   
667878741721415682 2015-11-21 01:34:35   
670046952931721218 2015-11-27 01:10:17   
671115716440031232 2015-11-29 23:57:10   
672594978741354496 2015-12-04 01:55:13   
700002074055016451 2016-02-17 17:01:14   
740365076218183684 2016-06-08 02:09:24   
745712589599014916 2016-06-22 20:18:30   
687818504314159109 2016-01-15 02:08:05   
669567591774625800 2015-11-25 17:25:28   
676821958043033607 2015-12-15 17:51:44   
887473957103951883 2017-07-19 00:47:34   
668643542311546881 2015-11-23 04:13:37   
796080075804475393 2016-11-08 20:00:55   
741303864243200000 2016-06-10 16:19:48   
671789708968640512 2015-12-01 20:35:22   
673708611235921920 2015-12-07 03:40:24   
674042553264685056 2015-12-08 01:47:22   
711306686208872448 2016-03-19 21:41:44   
716802964044845056 2016-04-04 01:41:58   
736225175608430592 2016-05-27 15:58:54   
738537504001953792 2016-06-03 01:07:16   
765222098633691136 2016-08-15 16:22:20   
774314403806253056 2016-09-09 18:31:54   
691820333922455552 2016-01-26 03:09:55   
...                                ...   
891689557279858688 2017-07-30 15:58:51   
667435689202614272 2015-11-19 20:14:03   
818536468981415936 2017-01-09 19:14:36   
674019345211760640 2015-12-08 00:15:09   
680206703334408192 2015-12-25 02:01:30   
685268753634967552 2016-01-08 01:16:17   
685547936038666240 2016-01-08 19:45:39   
680440374763077632 2015-12-25 17:30:01   
825876512159186944 2017-01-30 01:21:19   
832397543355072512 2017-02-17 01:13:34   
816336735214911488 2017-01-03 17:33:39   
719332531645071360 2016-04-11 01:13:34   
718246886998687744 2016-04-08 01:19:36   
686003207160610816 2016-01-10 01:54:44   
691459709405118465 2016-01-25 03:16:56   
672640509974827008 2015-12-04 04:56:09   
713411074226274305 2016-03-25 17:03:49   
748307329658011649 2016-06-30 00:09:04   
819347104292290561 2017-01-12 00:55:47   
827653905312006145 2017-02-03 23:04:02   
828372645993398273 2017-02-05 22:40:03   
699413908797464576 2016-02-16 02:04:04   
828409743546925057 2017-02-06 01:07:28   
697881462549430272 2016-02-11 20:34:41   
709566166965075968 2016-03-15 02:25:31   
689661964914655233 2016-01-20 04:13:20   
669214165781868544 2015-11-24 18:01:05   
834209720923721728 2017-02-22 01:14:30   
673317986296586240 2015-12-06 01:48:12   
674790488185167872 2015-12-10 03:19:24   

                                                                                                                                                                         text  \
tweet_id                                                                                                                                                                        
666051853826850816  This is an odd dog. Hard on the outside but loving on the inside. Petting still fun. Doesn't play catch well. 2/10 https://t.co/v5A4vzSDdc                  
666049248165822465  Here we have a 1949 1st generation vulpix. Enjoys sweat tea and Fox News. Cannot be phased. 5/10 https://t.co/4B7cOc1EDq                                    
666094000022159362  This appears to be a Mongolian Presbyterian mix. Very tired. Tongue slip confirmed. 9/10 would lie down with https://t.co/mnioXo3IfP                        
666373753744588802  Those are sunglasses and a jean jacket. 11/10 dog cool af https://t.co/uHXrPkUEyl                                                                           
666055525042405380  Here is a Siberian heavily armored polar bear mix. Strong owner. 10/10 I would do unspeakable things to pet this dog https://t.co/rdivxLiqEt                
667090893657276420  This is Clybe. He is an Anemone Valdez. One ear works. Can look in 2 different directions at once. Tongue slip. 7/10 https:

## Store

In [ ]:
twitter_archive.to_csv